### PR TITLE
Pull request for code review

### PR DESCRIPTION
**Pull request by MatthewBerglund**
**Project requirements:**
[X] It should be horizontally centered on the page.
[X] It should display a random quote when the page loads.
[X] It should display the name of the author of the quote when the page loads.
[X] When the "New Quote" button is clicked, it should fetch a new quote and display it.
[X] When the "New Quote" button is clicked, it should fetch the new quote's author and display it.
[X] It should provide a way for the user to tweet the current quote.

### CODE REVIEW

File: Requirements.txt [hunk 1]
```diff
@@ -1,8 +1,8 @@
 Random Quote Machine requirements:
```
Comment (motine) [inline Requirements.txt]: first, I must say this is a really neat project. It is fun and nicely done.
Below you will find my comments. I will be very detailed and "nit-picky". this does not mean that your work is bad or wrong, it just means that I may have an more detailed/different opinion on certain thing. please feel free to ignore comments and to resolve others in the merge request. I will add one comment per topic, so we can have individual conversation threads for each. but now, enough prelude...

File: Requirements.txt [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
 
-[] It should be horizontally centered on the page.
```
Comment (motine) [inline Requirements.txt:3]: The README is empty. why not use the readme for the requirements. I always find it really nice to enter a project on Github and see some intro, maybe a screenshot or similar.

suggestion: 1h research on "what makes a good readme"

File: Requirements.txt [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
-[] It should display the name of the author of the quote when the page loads.
-[] When the "New Quote" button is clicked, it should fetch a new quote and display it.
-[] When the "New Quote" button is clicked, it should fetch the new quote's author and display it.
-[] It should provide a way for the user to tweet the current quote.
+[X] It should be horizontally centered on the page.
+[X] It should display a random quote when the page loads.
+[X] It should display the name of the author of the quote when the page loads.
+[X] When the "New Quote" button is clicked, it should fetch a new quote and display it.
+[X] When the "New Quote" button is clicked, it should fetch the new quote's author and display it.
+[X] It should provide a way for the user to tweet the current quote.
```
Comment (motine) [inline Requirements.txt]: nice to include `[x]` here to see what is done and what is still outstanding.

maybe checkout [github's markdown](https://www.w3schools.io/file/markdown-checkbox-github/) for checkboxes.

File: index.html [hunk 4]
```diff
@@ -26,4 +26,4 @@
   </div>
```
Comment (motine) [inline index.html]: do you know what preconnect does?

suggested read on the basics of HTTP/IP
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Overview
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Session
- https://www3.ntu.edu.sg/home/ehchua/programming/webprogramming/HTTP_Basics.html (only skim, because it is complicated)

Comment (motine) [inline index.html]: you are only using divs. this is completely ok. however, in recent times it is preferable to use more "semantic" tags (e.g. main).

strongly recommended read:
- https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML

please consider reworking your html

File: index.html [hunk 5]
```diff
@@ -26,4 +26,4 @@
...
   <script src="main.js"></script>
```
Comment (motine) [inline index.html:31]: please explain: why do we put the script tag at the end?
(it is correct, but why?)

File: index.html (hunk 4 above)
Comment (motine) [inline index.html]: an you make the attribution link above open in another window?

File: main.js [hunk 6]
```diff
@@ -55,4 +55,4 @@ function updateTwitterIntent(quote) {
   var intentAddress = 'https://twitter.com/intent/tweet?text=';
```
Comment (motine) [inline main.js]: nicely, structured quotes in line 1 (github does not allow to put comments up there...)

Comment (motine) [inline main.js]: the use of `var` is "old" javascript. it works fine, it is safe to use ES6 nowadays.

strong recommendation for 1-2 days:
- go through course: https://es6.io/ (valentina should still have access to the course from mobileJob days; if not I will share mine)

consider using `const`, `let` and `() => {}`

Comment (motine) [inline main.js]: above you are using `animate`. super cool, I did not know this! Thanks for showing me.

Can you tell me how well the `animate` method is supported (use caniuse.com)? do you think using it is ok? what are factors to decide?

Comment (motine) [inline main.js]: a) what does `Math.ceil(0)` do?
b) why/when do you think it makes sense to use `Math.floor(...)?

File: main.js [hunk 7]
```diff
@@ -55,4 +55,4 @@ function updateTwitterIntent(quote) {
...
   intentAddress += '\"' + quote['text'] + '\"' + '  — ' + quote['movie'];
   document.getElementById('tweet-quote-link').setAttribute('href', intentAddress);
```
Comment (motine) [inline main.js]: nice touch with the twitter link!

in the ES6 course we see [Template Literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals). I think they would be perfect for this use case.

File: main.js [hunk 8]
```diff
@@ -55,4 +55,4 @@ function updateTwitterIntent(quote) {
   var intentAddress = 'https://twitter.com/intent/tweet?text=';
   intentAddress += '\"' + quote['text'] + '\"' + '  — ' + quote['movie'];
```
Comment (motine) [inline main.js]: why is there the same file in the assets directory?

Before publishing something (= every commit is a publishing act), I always review my work.

File: main.js [hunk 9]
```diff
@@ -55,4 +55,4 @@ function updateTwitterIntent(quote) {
...
   document.getElementById('tweet-quote-link').setAttribute('href', intentAddress);
-}
```
Comment (motine) [inline main.js:58]: nice work on the styles. they work well visually and are well structured.

File: main.js [hunk 10]
```diff
@@ -55,4 +55,4 @@ function updateTwitterIntent(quote) {
...
\ No newline at end of file
+}
```
Comment (motine) [inline main.js:60]: nice work on the title of the commits. it is really nice that you take the time to write nice captions.
don't be afraid to commit more often to avoid unrelated content (e.g. `Update button styles` also includes changes to the quote data. I often do multiple commits after a programming stretch.

**Review by motine — CHANGES_REQUESTED** [review]

File: Requirements.txt (hunk 1 above)
Comment (MatthewBerglund) [reply Requirements.txt]: Wow, that was fast. Thanks a lot for the comments! I'm gonna start working through them right away, and I'll request a new review when I'm done.

Just a heads up: I will be away on Friday and Saturday of this week, with occassional access to email.

File: index.html (hunk 4 above)
Comment (MatthewBerglund) [reply index.html]: Had to look it up. It seems that preconnect tells the browser to connect to the external resource in advance (in this case Google fonts) so that whatever is at the resource (a stylesheet with my chosen fonts? the fonts themselves?) can be fetched more quickly. I tried clearing my cache and commenting out lines 7 and 8 (the two `<link ref="preconnect">` elements) and it doesn't seem to make a difference.

Comment (motine) [reply index.html]: good.
maybe check this one out (it's ok if you only understand half of it): https://stackoverflow.com/a/47452035

Now that we know what it does. Is it worth keeping around?
Remember: every superfluous line means unnecessary complexity and maintenance work.

Comment (MatthewBerglund) [reply index.html]: I would say it isn't necessary in this case, but I'm also wondering: if it were worth keeping despite the added complexity, how would I know? Is it a matter of how quickly the page loads?

Comment (motine) [reply index.html]: yes, this `preconnect` thing helps loading the page faster. and here we see a core struggle in development: simplicity vs. performance. Often performance optimizations increase the complexity and worsen the maintainability. So it is always a trade off. In this case we can ask:
- how much of a burden is it to maintain this line of code up there? probably not much
- how much does the navigability (you know what i mean) suffer from these lines? well not too much either
- how much performance gain do we have? well, probably also not much.

in conclusion, here it really does not matter too much and you can choose freely. personally, I would remove it because I am strictly "less code is better code". but in your situation i would keep it, just to show other readers that you know what you are doing.

thanks for the great discussion. do you want to summarize what you have taken away from it? (consolidates the learning)

**Comment by motine** [comment]
@MatthewBerglund, one more thing. Have you thought about publishing this project via [github pages](https://pages.github.com/)?

File: index.html (hunk 4 above)
Comment (MatthewBerglund) [reply index.html]: Okay, so this lines up pretty nicely with the advice I've gotten so far with regard to programming: the first priorty should be to write understandable code, then we can consider performance loss/gain.

Since `preconnect` is just connecting to an external resource in advance so that the requested resource can be fetched faster, and since it doesn't seem to improve performance much in this case, it might be worth getting rid of to make the markup more readable/maintainable.

You say that I should keep it because it shows that I know what I'm doing, but all the Google Fonts markup (lines 7-9) was generated automatically by Google — all I did was copy and paste it into my HTML. I wonder if it really says anything at all about my knowledge. Or do you think removing it would make it look like I "forgot" the rest of the markup?

Comment (MatthewBerglund) [reply index.html]: Done. I think the HTML makes more sense semantically now. Resolving.

File: main.js (hunk 8 above)
Comment (MatthewBerglund) [reply main.js]: Good point. I was experimenting with two versions of the twitter icon and forgot to remove the one I didn't end up using. It's gone now.

File: index.html (hunk 5 above)
Comment (MatthewBerglund) [reply index.html:31]: I suppose for main.js we want the DOM to be there first so that the JavaScript can find the elements it needs to bind events and do DOM manipulation.

File: index.html (hunk 4 above)
Comment (motine) [reply index.html]: well, good point about the forgetting. I am undecided. Roll a dice :)

File: Requirements.txt (hunk 1 above)
Comment (motine) [reply Requirements.txt]: no worries, I have no stress :)

You concentrate on becoming a married man :)

File: main.js (hunk 6 above)
Comment (MatthewBerglund) [reply main.js]: Going through the ES6 course you linked. Imagine my surprise to learn that there are different schools of thought on if/when/how to use `let`, `const` and `var`. Man, using `var` for everything was such a breeze! 

I'm leaning towards using only `let` and `const`: using `let` by default, unless I know for sure that the value of the variable will not be reassigned, and then possibly refactoring variables to `const` at the end if it makes sense. The mental overhead of yet _another_ decision to make feels very "heavy" right now.

My thinking is that only having two variables to choose between lowers the mental overhead, and defaulting to `let` when unsure relieves me of having to know for certain whether the variable will get reassigned.

What do you think?

Comment (motine) [reply main.js]: well, hard to say. When I declare a variable I usually already know if it will change or not. However, I can see that this might be difficult sometimes when you are exploring different solutions. I would still flip it around. use `const` as long as you can. if it crashes then go to `let`. reasoning: `let` always works and is weaker, `const` expresses an intention and is stronger.

Comment (MatthewBerglund) [reply main.js]: Okay, that makes sense. `const` seems to work just about everywhere in this application.

Comment (MatthewBerglund) [reply main.js]: Hmm, I'm a little unsure how to interpret the support scores on caniuse.com. Maybe "89% supported" isn't great for production code, but for what I'm doing it's probably okay. The only people I expect to look at this are most likely going to be using modern browsers. If I were trying to make this available to a wider audience I would probably avoid using it. 

BUT, I personally see a lot of value in "pretending" to work under these types of constraints, so I think I would prefer to explore better-supported options.

I suppose the other thing to consider is whether this aspect of the app is essential, which it isn't. But if `.animate()` isn't supported on the user's browser, I think the quotes won't render correctly, and the app will actually look kind of "broken", so that's not great.

I think I'm going to try to implement the animation in a better way.

Comment (motine) [reply main.js]: ❤️

Comment (MatthewBerglund) [reply main.js]: Yeah, this section was not very well thought out.

a) `Math.ceil(0)` just returns 0, so it doesn't make any sense to use it.
b) You use `Math.floor()` when you want to round a floating point number down to the closest integer. Using it on `array.length` doesn't make sense because `array.length` will in any case be equal to an integer. 

It does make sense, however, to use `Math.floor()` on the value returned by `Math.random() * array.length` because that value will be a floating point number, and we ultimately want to round down so that we get an integer that excludes `array.length` from the result.

In the end, we don't need all the `(max - min) + min` stuff either because `min` is always 0. So we end up with: `Math.floor(Math.random() * array.length)`

**Comment by MatthewBerglund** [comment]
@motine Hey Tom, all finished making changes. Some of my commit messages don't make sense because I was using backticks `` inside the messages without realizing that that they serve a very specific purpose in git. So you might see some commits where words are missing.

File: index.html (hunk 5 above)
Comment (motine) [reply index.html:31]: cool beans
